File: DiscordConnector/PublicAPI/api/v0/channel/create.py
from fastapi import APIRouter, Depends
from DiscordConnector.PublicAPI.auth import require_admin
from DiscordConnector.PublicAPI.dependencies import get_channel_service
from ..schemas import ChannelCreate, ChannelResponse

router = APIRouter()


@router.post("/create", response_model=ChannelResponse)
async def create_channel(
    data: ChannelCreate,
    _principal=Depends(require_admin),
    service=Depends(get_channel_service),
):
    category_id = int(data.category_id) if data.category_id is not None else None
    channel = await service.create_channel(data.name, category_id, data.position)
    return ChannelResponse(
        id=str(channel.id),
        name=channel.name,
        category_id=(
            str(channel.category_id) if channel.category_id is not None else None
        ),
        position=channel.position,
    )

File: DiscordConnector/src/PublicAPI/api/v0/channel/create.ts
import { createRoute, type OpenAPIHono } from "@hono/zod-openapi";
import { channelResponse } from "../../../responses";
import {
  bearerSecurity,
  ChannelSchema,
  commonErrorResponses,
  createChannelBodySchema,
  getServices,
  jsonBody,
  jsonContent,
  requireRoleMiddleware,
  type AppEnv,
} from "../../../shared";

export function registerCreateChannelRoute(app: OpenAPIHono<AppEnv>): void {
  app.openapi(
    createRoute({
      method: "post",
      path: "/api/v0/channel/create",
      security: bearerSecurity,
      middleware: requireRoleMiddleware("admin"),
      request: {
        body: jsonBody(createChannelBodySchema, "Channel creation payload"),
      },
      responses: {
        200: jsonContent(ChannelSchema, "Created channel"),
        ...commonErrorResponses(),
      },
    }),
    async (c) => {
      const body = c.req.valid("json");
      const channel = await getServices(c.env).channelService.createChannel(
        body.name,
        body.category_id ?? null,
        body.position ?? null,
      );
      return c.json(channelResponse(channel), 200);
    },
  );
}
